# Classification

## K-Nearest Neighbour

Importing the required Python Packages

In [3]:
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import matplotlib.ticker as ticker
from sklearn import preprocessing
%matplotlib inline

Lets download the dataset. To download the data, we will use !wget to download it from IBM Object Storage.

In [4]:
!wget -O teleCust1000t.csv https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/teleCust1000t.csv

--2019-08-08 17:43:26--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/teleCust1000t.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.193
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.193|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 37048 (36K) [text/csv]
Saving to: ‘teleCust1000t.csv’

teleCust1000t.csv   100%[===================>]  36.18K   123KB/s    in 0.3s    

2019-08-08 17:43:28 (123 KB/s) - ‘teleCust1000t.csv’ saved [37048/37048]



**Creating a dataframe from the csv file**

In [5]:
df = pd.read_csv('teleCust1000t.csv')
df

region  tenure  age  marital  address  income  ed  employ  retire  \
0         2      13   44        1        9    64.0   4       5     0.0   
1         3      11   33        1        7   136.0   5       5     0.0   
2         3      68   52        1       24   116.0   1      29     0.0   
3         2      33   33        0       12    33.0   2       0     0.0   
4         2      23   30        1        9    30.0   1       2     0.0   
5         2      41   39        0       17    78.0   2      16     0.0   
6         3      45   22        1        2    19.0   2       4     0.0   
7         2      38   35        0        5    76.0   2      10     0.0   
8         3      45   59        1        7   166.0   4      31     0.0   
9         1      68   41        1       21    72.0   1      22     0.0   
10        2       5   33        0       10   125.0   4       5     0.0   
11        3       7   35        0       14    80.0   2      15     0.0   
12        1      41   38        1        8    37.0   2       9     0.0   
13        2      57   54        1       30   115.0   4      23     0.0   
14        2       9   46        0        3    25.0   1       8     0.0   
15        1      29   38        1       12    75.0   5       1     0.0   
16        3      60   57        0       38   162.0   2      30     0.0   
17        3      34   48        0        3    49.0   2       6     0.0   
18        2       1   24        0        3    20.0   1       3     0.0   
19        1      26   29        1        3    77.0   4       2     0.0   
20        3       6   30        0        7    16.0   3       1     0.0   
21        1      68   52        1       17   120.0   1      24     0.0   
22        3      53   33        0       10   101.0   5       4     0.0   
23        3      55   48        1       19    67.0   1      25     0.0   
24        3      14   43        1       18    36.0   1       5     0.0   
25        2       1   21        0        0    33.0   2       0     0.0   
26        2      42   40        0        7    37.0   2       8     0.0   
27        3      25   33        1       11    31.0   1       5     0.0   
28        1       9   21        1        1    17.0   2       2     0.0   
29        2      13   33        1        9    19.0   4       0     0.0   
..      ...     ...  ...      ...      ...     ...  ..     ...     ...   
970       3      48   45        0       15    41.0   1      10     0.0   
971       2       9   53        1       20    50.0   2      13     0.0   
972       1      61   52        0       21    82.0   1      18     0.0   
973       3      72   53        0       34   108.0   1      23     0.0   
974       3      57   60        0       20    14.0   2      27     1.0   
975       2      57   51        0       10    46.0   1      13     0.0   
976       1      37   60        1        0    46.0   1      20     0.0   
977       3      45   51        0       13   146.0   4      16     0.0   
978       3      45   39        1       15    36.0   1      11     0.0   
979       2      55   44        0       24    83.0   1      23     0.0   
980       2      24   27        0        2    17.0   3       3     0.0   
981       3       1   20        0        1    18.0   3       0     0.0   
982       2      34   23        1        3    24.0   1       7     0.0   
983       3      14   37        1        5    48.0   3       1     0.0   
984       3       6   32        0       10    47.0   1      10     0.0   
985       3      24   30        0        0    25.0   4       5     0.0   
986       2       3   26        1        6    59.0   4       0     0.0   
987       1       4   30        0        1    45.0   4       6     0.0   
988       1      72   40        1       19   163.0   4      15     0.0   
989       1      44   56        0        4    63.0   2       6     0.0   
990       1      50   43        0        6    27.0   3       4     0.0   
991       2      61   50        1       16   190.0   2      22     0.0   
992      

<b> About the dataset </b>

Imagine a telecommunications provider has segmented its customer base by service usage patterns, categorizing the customers into four groups. If demographic data can be used to predict group membership, the company can customize offers for individual prospective customers. It is a classification problem. That is, given the dataset,  with predefined labels, we need to build a model to be used to predict class of a new or unknown case. 

The example focuses on using demographic data, such as region, age, and marital, to predict usage patterns. 

The target field, called __custcat__, has four possible values that correspond to the four customer groups, as follows:
  1- Basic Service
  2- E-Service
  3- Plus Service
  4- Total Service

Our objective is to build a classifier, to predict the class of unknown cases. We will use a specific type of classification called K nearest neighbour.


In [6]:
df.head()

region  tenure  age  marital  address  income  ed  employ  retire  gender  \
0       2      13   44        1        9    64.0   4       5     0.0       0   
1       3      11   33        1        7   136.0   5       5     0.0       0   
2       3      68   52        1       24   116.0   1      29     0.0       1   
3       2      33   33        0       12    33.0   2       0     0.0       1   
4       2      23   30        1        9    30.0   1       2     0.0       0   

   reside  custcat  
0       2        1  
1       6        4  
2       2        3  
3       1        1  
4       4        3

Searching the presence of NaN values in every column

In [7]:
for col in df.columns:
    print('The no. of NaN values in column {} is : {}'.format(col,df[col].isnull().sum()))

The no. of NaN values in column region is : 0
The no. of NaN values in column tenure is : 0
The no. of NaN values in column age is : 0
The no. of NaN values in column marital is : 0
The no. of NaN values in column address is : 0
The no. of NaN values in column income is : 0
The no. of NaN values in column ed is : 0
The no. of NaN values in column employ is : 0
The no. of NaN values in column retire is : 0
The no. of NaN values in column gender is : 0
The no. of NaN values in column reside is : 0
The no. of NaN values in column custcat is : 0


<b> Lets see how many of each class is present in the cutsat column </b>

In [8]:
df['custcat'].value_counts()

3    281
1    266
4    236
2    217
Name: custcat, dtype: int64

<b> So we have :
    266 Basic Service
    217 E-Service
    281 Plus Service
    236 Total Service
</b